src/server/active_tran_server.cpp

```diff
@@ -202,6 +202,16 @@ active_tran_server::connection_handler::on_disconnecting ()
     }
 }
 
+void
+active_tran_server::connection_handler::prior_sender_sink_hook (std::string &&message)
+{
+  assert (message.size () > 0);
```

Comment:
Assert that the state is at least `CONNECTING`?
I know it is by design as long as this handler is only registered in `on_connecting`.

Reply:
It's registered on CONNECTING, but this function is called whenever a prior list is sent on `send_list`. Do I get it wrong?

Reply:
I know. That's why I said that it is "by design". It is registered when the state is `CONNECTING` and it executes:
1. during `CONNECTING` - after registration
2. during `CONNECTED` - the whole duration of the state
3. during `DISCONNECTING` - for a very short while as seen in function `connection_handler::disconnect_async`:
```
  assert (m_state == state::CONNECTING || m_state == state::CONNECTED);
  m_state = state::DISCONNECTING; // <<=================== (1)

  m_disconn_future = std::async (std::launch::async, [this, with_disc_msg]
  {
    on_disconnecting (); // server-type specific jobs before disconnect // <<==================== (2)
```
Between (1) and (2) - where the handler is unregistered log prior can still be sent.

My suggestion was to just add an assert to formalize these inter-dependencies towards the state.
```suggestion
  assert (m_state >= state::CONNECTING && m_state <= state::DISCONNECTING);
```
But now, looking back at it, the assert will be reading the state unguarded by mutex. I guess it's fine being only debug code.

I leave it up to you.

Reply:
Thank you for elaborating and I've got it. In this case, it'd be just for debugging but should open the m_state to derived classes. I leave it as it is, but I'll keep in mind the formalization with an assertion.